WIP adding the DLM (and LINKIMAGE) functionality.

GillesDuvert opened the pull request:
With this PR, the user has access to a LINKIMAGE and all DLM-related functionalities with DLL (.so, .dll) files compiled for IDL (no need to recompile) or (and this would be way better in terms of efficiency) directly using GDL internal code.

DLMs found in !DLM_PATH will be charged automatically by GDL, whatever their 'flavor' if the `.dlm` is correctly written. 

HELP,/DLM works. Also DLM_REGISTER.

###  IDL-compatible libraries
user: nothing to do. just use IDL_DLM_PATH to  point to the dlm files or use DLM_REGISTER. Example: fmedian in the NASA ssw suite (**however see footnote**, there is no gain in using fmedian)
programmer: Just follow IDL documentation.  The distribution comprizes the idl_export.h needed. 

###  Libraries directly compatible with GDL (no lengthy changes from C++ GDL variables to IDL C variables and back)
These have to be compiled with a C++ compiler, and must `#include "envt.hpp"`. Each function/procedure to be integrated withing GDL should just be defined as:
```
extern "C" BaseGDL* my_fun( EnvT* e) //function: gives back a gdl variable
extern "C" void my_pro( EnvT* e) //defines a procedure
```
my_pro and my_fun probably need to be **uppercase** 

**HOWEVER:** if the code contains its _own C++ templates_  then there are chances that some template specialization will not be present in the .so (.dll) and the library cannot be linked with GDL. TBC.

To illustrate, POWELL is now a DLM, in `gnudatalanguage/dlm `

### usage
To differentiate the two 'flavours', GDL-native .so (.dll) code must be loaded wih a minimal change wrt. the IDL fashion:

1. for DLMs: the DLM shoud **start** with a line saying `#%GDL_DLM`
2. for LINKIMAGE: add the /NATIVE keyword when calling LINKIMAGE.

### footnote
1. ssw'w fmedian.dlm does not seem faster than IDL's own on my machine (but: fmedian permits non-square "kernels"):
```
IDL> z=dist(512) & tic & a=fmedian(z,17,17) & toc
% Time elapsed: 0.60005403 seconds.
IDL> tic & a=median(z,17) & toc    
% Time elapsed: 0.31774592 seconds.
```
besides, fmedian and median do not produce the same result!!!

2. but both are way slower than GDL's MEDIAN() that should be preferred:
```
GDL> z=dist(512) & tic & b=median(z,17) & toc
% Time elapsed : 0.079225063 seconds.
```

 And yes, GDL's MEDIAN handles non-square kernels provided the x and y widths are odd:
```
GDL> z=dist(512) & tic & b=median(z,[11,17]) & toc
% Time elapsed : 0.050585985 seconds.
```
But use with caution as this is not IDL-compatible.

Commit:
progress!!

Commit:
progress!!

Commit:
pointer C problem

Commit:
progress

Commit:
keyword passing mechanism approximately secure - Arrays not completely supported, Structures absent.

Commit:
well well

Commit:
passes keyword_demo, mult2, rsum and fmedian.
Many functions still missing but this is in pretty good shape.
Only problem will be the uncertain memory loss.

Commit:
bug corrected - unlinkimage tests ok?

Commit:
artificially make all dll calls appear as different DlibPro (or DlibFun) when they are just 2 functions, CallDllPro and CallDllFunc. This to avoid multiple deletions in case unlinkimage  and unregistersymbol (GDL only) are to be kept.

Commit:
insure correct message (procedure not found) instead of 'variable not found' when implied_print doe not work... because it was a 'procedure not found'

Commit:
added dlm_load (do nothing at the moment, as all dlms are loaded using linkimage for the time being)

Commit:
added get_dlm, an overkill procedure to load all dlms at once or one in particular.

Commit:
use gdl_get_dlm at start, plus side changes

Commit:
correct

Commit:
create dlm install directory

Commit:
dlm_register must not up- or lowcase the function/procedure name written in the DLL.

Commit:
to correctly use DLM; <IDL_XX> default values must be correctly handled. So a dozen of files had to be changed.

Commit:
update median to take advantage of 'rectangular' kernels allowed by some of our fast codes. This is a non-IDL feature impossible to test against IDL results for correctness.

Commit:
String should use snprintf!

Commit:
apple clang says: format specifies type 'unsigned int' but the argument has type 'size_t' (aka 'unsigned long') [-Wformat] and suggest format '%zu'. Trying it.

Commit:
test RTD_LOCAL

Commit:
corrected 2 bugs

Commit:
Use POWELL (brent.cpp) as an example for GDL "native" DLL.

Commit:
PFFF... CMake sucks.

Commit:
no replication of % Loaded DLM message (correct call of LINKIMAGE in DLM_REGISTER.PRO).
Handles uppercase names of functions and mixed-case name of corresponding entries in the .so or .dll

Commit:
valgrind found double deletion at exit, removed culprit.

Commit:
powell exemple in lowercase (more normal for a c function and better example for the dlm)

Commit:
added a bunch of functions

Commit:
added a bunch of functions, corrected some errors.

Commit:
dlm_register must find the dlms in the current directory at start.

Commit:
try a better approach to the problem of releasing memory correctly.

Commit:
still clearer and less error-prone. Also solved logical error in IDL_KWProcessByOffset and hopefull IDL_KWGetParams

Commit:
better yet

alaingdl commented:
I can compile with (apparent) success on Debian 10
`% Loaded DLM: POWELL`

But we cannot compile on OSX (large update of Xcode & LLVM last days). We will try hard !

GillesDuvert commented:
Well, thanks, that's good news for a start. I'm slowly polishing (!) the thing. (IDL<->GDL Structures are tough.)
Not soon in production, I'm afraid...

liufrancois commented:
Hi, I'm trying to build the PR on a Mac Mini M2 running macOS 15.5, but I ran into a few issues:

when compiling with Clang(20.1.5):
`make -j` stops at around 30% with no error message.

In `src/export.hpp`, I got an error: `'ulong' was not declared`. I fixed it by adding `typedef unsigned long ulong;`.

In `src/dlm/CMakeLists.txt`, I had to change `brent.cpp` to `brent.hpp` to remove an error.

In `src/basic_pro_jmg.cpp`, I had to remove `RTLD_DEEPBIND` to avoid a compilation error.

I also encountered an error with an `unknown option: --dynamic-list=.../src/dynlist.txt`.


I also tried compiling with GCC(11.5.0), and to reach the same point successfully, I had to disable several components by adding these flags:
`-DX11=off -DWXWIDGETS=off -DPYTHON=off -DREADLINE=off`

Hope this helps with improving compatibility on macOS. Let me know if you'd like me to test anything else.

GillesDuvert commented:
@liufrancois thanks, I'm aware of issues on OSX and Windows. Quite normal in fact. Will attend these issues once the linux version is OK, which it is not. At that moment the "WIP" in the PR title will disappear. Thanks again.

Commit:
structures 1/2 made and problem in dlm_register solved.

Commit:
found some problems -- just a safeguard commit

GillesDuvert commented:
@liufrancois if the goal is to get a gdl-only dlm to work, aside the problems you solved,  you need to find how to get all symbols in the gdl executable available to the "loaded" .so file. This is  the purpose of the link option that osx/clang does not seem to accept.

Commit:
structures OK, but inherited structures...

Commit:
corrected a regression in gdl_implied_print.pro

Commit:
alignment in structs correctly computed.

Commit:
alignment in structs NOW correctly computed!!
(and alignment in autoprint, too.)

Commit:
Well, alignment in structs NOW correctly computed!!!!!

Commit:
added a bunch of functions

Commit:
added various examples of idl-type dlm in testmodule.c

Commit:
corrected label_date (#2037)
made WIDGET_INFO warns visible.

Commit:
added correction suggested by @liufrancois
tentatively put commands for OSX to export the symbols in the executable ( suggested by stackoverflow )

Commit:
will that pass ?

Commit:
test OSX

Commit:
test OSX

Commit:
IDL_** symbols must not be inline, silly!

Commit:
test again

Commit:
test something for rebuild windows.

Commit:
remove test_hdf5 which fails everywhere.

Commit:
test something for rebuild windows (2)

Commit:
waiting for a better solution, this permits the GDL-native DLMs on OSX.

Commit:
replace wxwidgets by wxwidgets3.3 for windows
remove test that fail on OSX

Commit:
replace wxwidgets by wxwidgets3.2 for windows

Commit:
remove test that fail (grrr!!!!)

Commit:
oops old code was not removed

Commit:
patch for OSX 13 ??

Commit:
patch for OSX 13 ??

Commit:
patch for OSX 13 ??

Commit:
remove graphicsmagick rebuild windows

Commit:
remove graphicsmagick from main.yml

Commit:
remove graphicsmagick from main.yml

Commit:
remove ??? from main.yml

Commit:
switch to portablexdr for windows...

Commit:
again, no portablexdr for windows??

Commit:
again, no portablexdr for windows??

Commit:
again, no portablexdr for windows??

Commit:
remove xdr specialization for WINDOWS.

Commit:
try something

Commit:
just remove longlong?

Commit:
prevent compiler warnings

Commit:
prevent compiler warnings (print -> nprint)

Commit:
prevent obsoleted "register"

Commit:
ugly undetected bug corrected

Commit:
ugly undetected bug corrected

Commit:
prevent compiler warnings (print -> nprint) etc.

Commit:
reverted erroneous patch triggered by this warning: implicit conversion from 'type' (aka 'unsigned long long') to 'DDouble' (aka 'double') changes value from 18446744073709551615 to 18446744073709551616 [-Wimplicit-const-int-float-conversion]

Commit:
tentative patches to Windows version of DLM-related stuff

Commit:
remove dprintf and pacify windows.

Commit:
adjust format length to IDL_LONG64

Commit:
xdr and windows.

Commit:
try something for xdr and windows: use our own xdr.h

Commit:
test for windows' portable xdr

Commit:
tags some absent IDL_CDECL

Commit:
do nothing hoping that DLL_PUBLIC will make the functions visible under windows. Exporting everything makes an error "export ordinal too large: 75852"

Commit:
use a DLL_PUBLIC tag for enabling visibilty of functions under windows.

Commit:
avoid for the moment on Windows.

Commit:
EXT_STDIO_FILEBUF_H must be tested only in 'cxx' mode.

Commit:
revert to the use of bsd-xdr, but patch it to enable compilation with recent mingw64 compiler.

Commit:
signal that UNIT keyword for SPAWN is potentially feasible on WINDOWS, just some cumbersome copy/paste to do.

Commit:
avoid a compiler warning about strncat

Commit:
remove DLL_EXPORT definition --- not used

Commit:
oops forgot to put back bsd-xdr cache

Commit:
oops forgot to put back bsd-xdr cache
